### ISSUE TITLE
Re-structure final YAML files

### ISSUE THREAD

**Issue by brynpickering** [calliope-project/euro-calliope#145]
As evidenced by recent PRs (#131, #139), there are many possible Calliope overrides being built in the workflow. These sometimes end up in their own file, sometimes in a file with technology definitions. We should consider re-structuring the YAML files with two key aims:

1. Modularisation. It should be possible to build the set of YAML files with any future combination of technologies/carriers in the system, without the content of the YAML files themselves changing much. 
2. Ease of tweaking in a pre-built model. There will be a set of users working directly with the pre-built model and never with this workflow. It should be intuitive for users to navigate the YAML files to update/add/remove components of the model as they see fit. This also fits for those elements that are not configurable in the Euro-Calliope Snakemake workflow, but a user could change once the model is built (e.g. in #139, a user might want to change from 4hrs being the threshold between the technology types).

Some thoughts on this:
1. Technologies and their overrides and allocation to locations could be packaged on a per-YAML file basis.
2. Overrides could be in a file of the same name as the override, all kept in an override directory.

**Comment by timtroendle** [comment]
I added this to 1.2 because this is a massive change and it gets worse the longer we wait.

**Comment by brynpickering** [comment]
This is what the file structure I propose in #212 looks like with the current Euro-Calliope:

![image](https://user-images.githubusercontent.com/17178478/151200548-752f48be-165e-4645-bc68-4cb15ab61fb8.png)

Each technology YAML file contains the technology definitions, any overrides, and the allocation of the technologies to locations, e.g.:
<details>
<summary>supply-hydro.yaml</summary>
<br>

```yaml
techs:
    hydro_run_of_river:  # from [@JRC:2014] Table 14
        essentials:
            name: Run of river hydro electricity
            parent: supply
            carrier: electricity
        constraints:
            resource: file=supply/capacityfactors-hydro-ror.csv
            resource_unit: energy_per_cap
            lifetime: 60
        costs.monetary:
            energy_cap: 562.0 # (10,000 EUR2013/MW)
            om_annual: 16.86 # (10,000 EUR2013/MW) 3% of CAPEX
            om_prod: 0.0005 # (10,000 EUR2013/MW)
    hydro_reservoir:  # from [@JRC:2014] Table 12
        essentials:
            name: Hydro electricity with a reservoir.
            parent: supply_plus
            carrier: electricity
        constraints:
            resource: file=supply/capacityfactors-hydro-reservoir-inflow.csv
            resource_unit: energy_per_cap
            lifetime: 60
        costs.monetary:
            energy_cap: 337.0 # (10,000 EUR2013/MW)
            om_annual: 10.110000000000001 # (10,000 EUR2013/MW) 3% of CAPEX
            om_prod: 0.0005 #  (10,000 EUR2013/MW)

locations:
    IRL.techs:
        hydro_run_of_river:
            constraints:
                energy_cap_max: 0.00216 # (100,000 MW)
        hydro_reservoir:
            constraints:
                energy_cap_max: 1e-05 # (100,000 MW)
                storage_cap_max: 0.009007527093596061 # (100,000 MWh)
    GBR.techs:
        hydro_run_of_river:
            constraints:
                energy_cap_max: 0.008484750000000003 # (100,000 MW)
        hydro_reservoir:
            constraints:
                energy_cap_max: 0.005240000000000001 # (100,000 MW)
                storage_cap_max: 4.719944197044335 # (100,000 MWh)

overrides:
    no-hydro-supply-fixed-cost:
        # Only assign costs to O&M and variable costs
        techs.hydro_run_of_river.costs.monetary:
            energy_cap: 0
            om_con: 0
        techs.hydro_reservoir.costs.monetary:
            energy_cap: 0
            storage_cap: 0
            om_con: 0

    schroeder-hydro-cost: # from [@schroeder:2013]
        techs.hydro_run_of_river:
            constraints.lifetime: 60
            costs.monetary:
                energy_cap: 300.0 # (10,000 EUR2010/MW) Table 33
                om_annual: 5.529779844 # (10,000 EUR2010/MW) Table 34
                om_prod: 0.0001 # (10,000 EUR2010/MW) stolen from om_annual for technical reasons
        techs.hydro_reservoir:
            constraints.lifetime: 60
            costs.monetary:
                energy_cap: 200.0 # (10,000 EUR2010/MW) Table 33
                om_annual: 1.529779844 # (10,000 EUR2010/MW) Table 34
                om_prod: 0.0001 # (10,000 EUR2010/MW) stolen from om_annual for technical reasons

    freeze-hydro-supply-capacities:
        locations:
            IRL.techs: # IRL
                hydro_run_of_river:
                    constraints:
                        energy_cap_equals: 0.00216 # (100,000 MW)
                hydro_reservoir:
                    constraints:
                        energy_cap_equals: 1e-05 # (100,000 MW)
                        storage_cap_equals: 0.009007527093596061 # (100,000 MWh)
            GBR.techs: # GBR
                hydro_run_of_river:
                    constraints:
                        energy_cap_equals: 0.008484750000000003 # (100,000 MW)
                hydro_reservoir:
                    constraints:
                        energy_cap_equals: 0.005240000000000001 # (100,000 MW)
                        storage_cap_equals: 4.719944197044335 # (100,000 MWh)
```
</details>

Thoughts on this structure @FLomb @sjpfenninger @timtroendle ?

**Comment by timtroendle** [comment]
This structure looks highly flexible and modular. It comes at the cost of being spread over a large amount of files, but I think the cost isn't too high.

When it comes to the question whether `techs` should be defined in separation from the locations (and therefore only once, instead of once for each spatial resolution) I have a slight preference for the current approach as it combines every definition at one place -- even if that means the tech definition is redundant.

I vote for a two minor changes: First, swapping the spatial resolution level and the current first-level folders. Instead of `./techs_and_locations/continental` it should be `./continental/techs_and_locations/` and `./continental/timeseries_data/` because then all resolution-specific things are at the same place. Second, for consistency, `techs_and_locations` should be `techs-and-locations`; `timeseries_data` should be `timeseries-data` (or even better: `timeseries`).

Apart from that, I guess we need to gain experience with this structure to fully understand its peculiarities.

**Comment by brynpickering** [comment]
1. this was a topic of conversation in the mini workshop we did internally. The issue with the resolutions being at the top level is that they are a bit abstract in the context of standard Calliope models. I am leaning towards the idea that we should go even further: we have three 'model' subdirectories in build, one for each resolution. If we're duplicating some tiny YAML content already (tech definitions), we can do the same for 'interest-rates.yaml' etc. This means that if you only care about one resolution, you can effectively remove all reference to the resolution in the file structure.
```
│── models
│   │── readme.md
│   │── continental   
│   │   ├── model.yaml
│   │   ├── scenarios.yaml     
│   │   ├── interest-rates.yaml 
│   │   ├── locations.yaml   
│   │   ├── techs            
│   │   └── timeseries
│   │── national        
...
```

2. sure thing.

3. It's difficult to get this right as the files include tech definitions _and_ the allocation of techs to locations. Nevertheless, your suggestion is fine by me.

**Comment by timtroendle** [comment]
1. I like your idea.

**Comment by sjpfenninger** [comment]
I think this looks good. The revised structure (1) is good, nothing from my side to complain about. With regards to (3) I am also not sure if any solution is ideal, but also happy with Tim's suggestion and maybe as Tim suggests, experience will show if some more minor tweaking makes sense later

**Comment by FLomb** [comment]
No strong preference about what should be at the top level of the folder structure. I would be perfectly fine with Bryn's latest proposal of having 3 model folders for the three resolutions and everything else in those, as long as the discussed improvements in the sub-folders remain.

As far as having a single location file with all techs versus having locations specified for each tech type in the corresponding tech.yaml, I think there is no win-win solution. As Tim said, if you have locations specified for each tech in independent files, you lose what's happening overall in a given location, and that's a very good point. But, on the other hand, having the existence of a given tech (say, chp) across all locations specified in the same file that characterises the tech would be very helpful for those that use only the pre-built models. It would allow for a much easier switching on/off of individual tech (groups), compared to the other case in which one would need to go in the huge overall locations.yaml and manually jump through a multitude of teachs across locations to switch things on/off - or, create an override that iterates across the needed locations for the needed techs. So I would say option A (one big locations.yaml) benefits the overall understanding of what's in each location; whilst option B (tech-specific locations in specific-tech.yaml files) makes easier the customisation of the pre-built model.

**Comment by brynpickering** [comment]
I agree with the relative merits of the two structures @FLomb. One option I floated with Tim was to have a top level file to describe all the technologies, so that it is easy to see all the techs that exist in the model. It would just be an 'existence' YAML, e.g.:

```yaml
techs:
    open_field_pv.exists: true
    roof_mounted_pv.exists: true
    wind_onshore_monopoly.exists: true
    wind_onshore_competing.exists: true
    wind_offshore.exists: true
    hydro_run_of_river.exists: true
    hydro_reservoir.exists: true
    biogas_to_electricity_supply.exists: true
    ....
``` 

Then a user can quickly see all the technologies that exist in the model and switch them on/off with ease, without having to delve into each of the more detailed YAML files. They would also know the names of all the techs, so they can search for them in the YAMLs if they want to change something about them.

**Comment by brynpickering** [comment]
How it now looks in #212:

<img width="230" alt="image" src="https://user-images.githubusercontent.com/17178478/152122189-ac0f39b0-b26b-4a34-8254-3e6b2fa70a08.png">

**Comment by sjpfenninger** [comment]
Seems good, just one nitpick related to the specific example: why `demand-electricity.yaml` but `electricity-demand.csv`?

**Comment by timtroendle** [comment]
I like the structure.

**Comment by brynpickering** [comment]
> Seems good, just one nitpick related to the specific example: why `demand-electricity.yaml` but `electricity-demand.csv`?

Agreed, it is odd. It is just legacy from the existing filenames. I was thinking of dealing with minor issues like this in a follow-up PR, independent of #212.

**Comment by FLomb** [comment]
I like the structure too
